docs/architecture/bignum.md

```diff
@@ -0,0 +1,129 @@
+                            size_t n, mbedtls_mpi_mod_modulus *N )
+
+int mbedtls_mpi_mod_mul( mbedtls_mpi_mod_residue *X, mbedtls_mpi_mod_residue *A,
+                         mbedtls_mpi_mod_residue *B, mbedtls_mpi_mod_modulus *N )
+
+int mbedtls_mpi_mod_add( mbedtls_mpi_mod_residue *X, mbedtls_mpi_mod_residue *A,
+                         mbedtls_mpi_mod_residue *B, mbedtls_mpi_mod_modulus *N )
+
+int mbedtls_mpi_mod_neg( mbedtls_mpi_mod_residue *X, mbedtls_mpi_mod_residue *A,
+                         mbedtls_mpi_mod_modulus *N )
+
+int mbedtls_mpi_mod_sub( mbedtls_mpi_mod_residue *X, mbedtls_mpi_mod_residue *A,
+                         mbedtls_mpi_mod_residue *B, mbedtls_mpi_mod_modulus *N )
+
+int mbedtls_mpi_mod_prime_inv( mbedtls_mpi_mod_residue *X,
+                               mbedtls_mpi_mod_residue *A,
+                               mbedtls_mpi_mod_modulus *P )
+```
+
+As mentioned above, there will be a second set of arithmetic functions with the same semantics. This second set of functions will be only used in the ECP module and won't perform any sanity checks on the parameters.
+
+There are several occasions where currently other operations are used, but these can be implemented with the above functions. (For example `mbedtls_mpi_sub_int_mod()` is only ever called on 3 and can be replaced by three calls to `mbedtls_mpi_mod_sub()`.)
+
+### Utility Functions
+
+There are a number of utility functions used throughout the ECC modules:
+
+```C
+mbedtls_mpi_get_bit()
+mbedtls_mpi_set_bit()
+mbedtls_mpi_copy()
+mbedtls_mpi_lset()
+mbedtls_mpi_cmp_int()
+mbedtls_mpi_cmp_mpi()
+mbedtls_mpi_safe_cond_assign()
+mbedtls_mpi_safe_cond_swap()
+mbedtls_mpi_random()
+mbedtls_mpi_fill_random()
+mbedtls_mpi_size()
+mbedtls_mpi_bitlen()
+```
+
+For these we need to provide equivalent functionality for the new type while respecting the pre and post conditions described above.
+
+Some of these might not be called directly in the end (for example `mbedtls_mpi_size()` is either used in I/O or can be inferred from the context.)
+
+### Input/Output
+
+We need to provide analogues to the standard input/output functions:
+
+```C
+mbedtls_mpi_write_binary_le()
+mbedtls_mpi_read_binary_le()
+mbedtls_mpi_write_binary()
+mbedtls_mpi_read_binary()
+```
+
+(`mbedtls_mpi_read_string()` is only ever used in selftest and in a function that is not called from anywhere and therefore can be ignored.)
+
+`derive_mpi` in ECDSA does some form of input as well and will need special considerations when switching to the new type.
+
+## Conversion
+
+Beyond the operations above we will need to convert to and from `mbedtls_mpi` as this is what public facing interfaces use. Since `mbedtls_mpi` and `mbedtls_mpi_mod_residue` use the same limb type, conversion might do copy or aliasing. To keep heap consumption low, we will be providing both and using the most suitable method based on the parameter type.
```

Comment:
We may have to distinguish conversion between public MPI and private MPI, as well as conversion between public and internal _format_ (concretely: Montgomery form). When converting from public to private MPI, we'll also always need a format conversion, but if e.g. we want to write an internal MPI, we'd only need a format conversion. 

One could try to hide the fact that an alternative presentation exists, but I'd find it rather unintuitive if, as a consequence, something like a binary write would need to perform non-trivial arithmetic first. I'd find it clearer if the internal modular arithmetic API exposed explicit conversion functions as in https://github.com/hanno-arm/mbedtls/blob/ecp_prototype/library/ecdsa.c#L426.

Reply:
The private and public distinction will definitely come up in RSA, but I am not sure about ECC. Do you have an example in mind?

Currently we don't use Montgomery form in ECC and we can still add format conversion later if we decide to do it explicitly instead of hiding the representation.

Our preferences differ in this regard: I think it is better to hide the representation and I very much would like to do so. That said, I can see some challenges and I am not sure if it will be practical.

I am happy to discuss/continue discussing this, but I don't think we need to make a decision on this urgently. (We will definitely need to make that decision before we start adding Montgomery arithmetic.)

Reply:
> Currently we don't use Montgomery form in ECC and we can still add format conversion later if we decide to do it explicitly instead of hiding the representation.

Ah, I was assuming that we will definitely switch to Montgomery in ECC immediately as part of the introduction of the new API -- simply because I didn't see any simple alternative so far. Was your understanding different, @yanesca? 

>Our preferences differ in this regard: I think it is better to hide the representation and I very much would like to do so. That said, I can see some challenges and I am not sure if it will be practical. 

Yes, in principle hiding irrelevant information simplifies things, but I think the API becomes a bit more complicated or at least unintuitive if one tries to hide the _existence_ of a non-trivial format conversion here, with the need for non-trivial arithmetic as part of say `write_binary()` being one example. 

Sounds like it would be good to at least agree if we think that the implementation of the proposed API for ECC will use Montgomery presentation or not? @mpg I seem to remember that you felt we'd be going for Montgomery, too, but I may misremember. Can you share your thoughts?

Reply:
> Sounds like it would be good to at least agree if we think that the implementation of the proposed API for ECC will use Montgomery presentation or not?

We discussed this and agreed that we would use Montgomery presentation.